feat: support for runtime paramters

scoheb opened the pull request:
- runtime parameters are sometimes used in resolver parameters.
- this PR adds the ability for users to specify runtime parameters on the command line.

scoheb committed:
feat: support for runtime paramters

- runtime parameters are sometimes used in resolver
  parameters.
- this PR adds the ability for users to specify runtime parameters
  on the command line.

Assisted-by: Cursor
Signed-off-by: Scott Hebert <scoheb@gmail.com>

lcarva commented on internal/validator/pipeline.go:
```diff
@@ -14,7 +15,39 @@ import (
+// resolveParameterExpressions resolves Tekton parameter expressions like $(params.paramName) with runtime values
+func resolveParameterExpressions(input string, runtimeParams map[string]string) string {
+	// Match $(params.paramName) pattern
+	re := regexp.MustCompile(`\$\(params\.([^)]+)\)`)
```
I think we can leverage pac to do this by adding the additional parameters to [this](https://github.com/konflux-ci/tektor/blob/db6caa2abda705df83d73d6b668ee85b7ff5e90f/internal/pac/pac.go#L61) map. That should make both `resolveParameterExpressions` and `resolveParamsInTaskRefParams` no longer needed.

scoheb replied:
removed

scoheb committed:
add gitignore for history files

lcarva commented on cmd/validate/validate.go:
```diff
@@ -58,7 +83,7 @@ func run(ctx context.Context, fname string) error {
 		}
 
-		if err := validator.ValidatePipelineRun(ctx, pr); err != nil {
+		if err := validator.ValidatePipelineRun(ctx, pr, runtimeParams); err != nil {
```
I don't think `ValidatePipelineRun` should care about the runtime params. Instead, this should be passed to the `pac.ResolvePipelineRun` call above. That's where the pac functionality is already added.

For Pipeline, instead of PipelineRun, we should have a similar pattern. Let's create a `pac.ValidatePipeline` to handle it. By the time we get to the validation functions, all pac tasks and parameters should've been resolved already.

scoheb replied:
updated.

scoheb committed:
add resolvepipeline to pac

- add ResolvePipeline to process both .tekton directory and the specific
pipeline file being validated, allowing validation of pipelines in
any directory structure.
- all paramters resolution now happens in the Resolve functions

Signed-off-by: Scott Hebert <scoheb@gmail.com>

lcarva commented on internal/pac/pac.go:
```diff
@@ -110,6 +117,9 @@ func ResolvePipelineRun(ctx context.Context, fname string, prName string) ([]byt
 	}
 
+	// Apply additional parameter substitutions to the PipelineRun structure
+	applyParameterSubstitutionsToPipelineRun(pr, params)
```
Why is this needed?

scoheb replied:
both **applyParameterSubstitutionsToPipelineRun** and **applyParameterSubstitutionsToPipeline** are needed because PaC only resolves variables with this pattern `{{ git-url }}` while Tekton resolves `$(params.git-url)`

Therefore:
- PaC's `templates.ReplacePlaceHoldersVariables` won't touch `$(params.git-url)` because it only looks for `{{git-url}}`
- Tekton's `substitution.ApplyReplacements` will replace `$(params.git-url)` with the actual value

lcarva replied:
But regular parameters should be resolved through the usual process. There's already code for handling that. I commented out this line, and the tests are still passing. Do you have an example of a pipeline run where this breaks?

lcarva replied:
Oh hang on. I think I know what you mean.

lcarva replied:
I get it now. This is needed to handle the case where the `taskRef` contains parameters. Failure to do so means that tektor will not be able to resolve the task reference.

That's not a PaC thing thing though.

I think it would make sense to have two sets of parameters. Ones that are pac variables and ones that are Pipeline parameters. Treating those two groups of parameters as one is likely going to be a problem for more sophisticated use cases.

I suggest the following:

1. Have two types of args for the validate command: `--param` and `--pac-param`. For Pipeline:
    - When calling `pac.ResolvePipeline`, pass in the values from `--pac-param`.
    - When calling `validator.ValidatePipeline` pass in the values from `--param`.

1. For PipelineRun:
    - When calling `pac.ResolvePipelineRun`, pass in the values from `--pac-param`.
    - When calling `validator.ValidatePipelineRun`, *don't* pass in the values from `--param`. (Log a warning that those are ignored for PipelineRun). The reasoning is that when a PipelineRun is created, it already has the parameters that it will run with. A user doesn't have a chance of changing the PipelineRun parameters without changing the PipelineRun itself.

1. The `ResolvePipeline*` functions should continue to use `templates.ReplacePlaceHoldersVariables`.

1. The `ValidatePipeline` function should take a new parameter that is a map of the pipeline parameters to be used during validation. I think you can move the call to `substitution.ApplyReplacements` to somewhere in `taskSpecFromPipelineTask`. We only want to do that for the `TaskRef.Params`.

1. The `ValidatePipelineRun` function should extract the params embedded in the PipelineRun, and pass those along to the `ValidatePipeline` function.

scoheb replied:
I believe I have covered all the points.

lcarva commented on internal/pac/pac.go:
```diff
@@ -120,6 +130,221 @@ func ResolvePipelineRun(ctx context.Context, fname string, prName string) ([]byt
+	}
+	templatePaths = append(templatePaths, fname)
+
+	allTemplates := templates.ReplacePlaceHoldersVariables(enumerateFiles(templatePaths), params)
```
Can we just return `[]byte(allTemplates)` here?

scoheb replied:
I think that since we still need to call `applyParameterSubstitutionsToPipeline(pipeline, params)`, we need to unmarsal to get our Pipeline (and PipelineRun) objects.

lcarva commented on internal/pac/pac.go:
```diff
@@ -120,6 +130,221 @@ func ResolvePipelineRun(ctx context.Context, fname string, prName string) ([]byt
 }
 
+func ResolvePipeline(ctx context.Context, fname string, pipelineName string, runtimeParams map[string]string) ([]byte, error) {
+	run := params.New()
```
This duplicates most of the stuff from ResolvePipelineRun. It's ok for now, but we should consider putting the common parts into a function.

scoheb replied:
created new setup function that is shared.

scoheb committed:
feat: support pac-params as well

- we now have pac-params for PaC parameter resolution
  and params for Tekton parameter resolution

Signed-off-by: Scott Hebert <scoheb@gmail.com>
